### PR TITLE
[MM-53269] Add configuration setting for integration requests timeout

### PR DESCRIPTION
**Pull request by hanzei**
#### Summary
Admins set up integrations to interact with external systems. These integrations make use of the MM toolkit, like slash commands and post actions. All of these toolkit elements have in common that they use HTTP requests to communicate with the integration.

The integration might take a long time to generate an HTTP response, e.g. an AI chat service can take minutes to answer (https://forum.mattermost.com/t/solved-outgoing-webhook-timeout-time/4013/4 ). Or the response is delayed due to latency (https://forum.mattermost.com/t/trigger-id-lifetime/8562/5 ).

To accommodate for these use cases, a new configuration setting is added that controls the timeout for all outgoing integration requests.  This includes:

- Outgoing webhooks
- Custom slash commands
- Post actions
- Interactive dialog requests (Through the trigger timeout)

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-53269

#### Screenshots
Screenshot from 2023-06-20 17-33-09

#### Release Note
```release-note
Add configuration setting for integration requests
```

### CODE REVIEW

File: webapp/channels/src/utils/constants.tsx [hunk 1]
```diff
@@ -1138,6 +1138,8 @@ export const DeveloperLinks = {
     ENABLE_OAUTH2: 'https://mattermost.com/pl/enable-oauth',
     INCOMING_WEBHOOKS: 'https://mattermost.com/pl/incoming-webhooks',
     OUTGOING_WEBHOOKS: 'https://mattermost.com/pl/outgoing-webhooks',
+    INTERACTIVE_MESSAGES: 'https://mattermost.com/pl/interactive-messages',
+    INTERACTIVE_DIALOGS: 'https://mattermost.com/pl/interactive-dialogs',
```
Comment (hanzei) [inline webapp/channels/src/utils/constants.tsx:1142]: These are already set up.

File: e2e-tests/playwright/support/server/default_config.ts [hunk 2]
```diff
@@ -91,6 +91,7 @@ const defaultServerConfig: AdminConfig = {
         EnableIncomingWebhooks: true,
         EnableOutgoingWebhooks: true,
         EnableCommands: true,
+        OutgoingIntegrationRequestsTimeout: 30,
```
Comment (isacikgoz) [inline e2e-tests/playwright/support/server/default_config.ts:94]: Maybe it's better to name it as `OutgoingIntegrationRequestsTimeoutSeconds`

File: server/channels/app/webhook_test.go [hunk 3]
```diff
@@ -810,16 +809,34 @@ func TestDoOutgoingWebhookRequest(t *testing.T) {
...
+		})
 
 		_, err := th.App.doOutgoingWebhookRequest(server.URL, strings.NewReader(""), "application/json")
 		require.Error(t, err)
 		require.IsType(t, &url.Error{}, err)
 	})
 
+	t.Run("with a slow response, long timeout configured", func(t *testing.T) {
+		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			time.Sleep(3 * time.Second)
```
Comment (isacikgoz) [inline server/channels/app/webhook_test.go:823]: Could we update this test and others test as `1` second sleep and `2` second timeout? Just to save some CI time.

File: server/i18n/en.json [hunk 4]
```diff
@@ -8009,7 +8009,7 @@
   },
   {
     "id": "interactive_message.decode_trigger_id.expired",
-    "translation": "Trigger ID for interactive dialog is expired. Trigger IDs live for a maximum of {{.Seconds}} seconds."
+    "translation": "Trigger ID for interactive dialog is expired. Trigger IDs live for a maximum of {{.Duration}}."
```
Comment (isacikgoz) [inline server/i18n/en.json:8012]: Is `Duration` translation friendly?

**Review by isacikgoz — COMMENTED** [review]
Overall LGTM, have some minor comments.

File: server/channels/app/slashcommands/command_test.go [hunk 5]
```diff
@@ -424,24 +423,42 @@ func TestDoCommandRequest(t *testing.T) {
...
+			io.Copy(w, strings.NewReader("Hello, World!"))
 		}))
 		defer server.Close()
 
-		th.App.HTTPService().(*httpservice.HTTPServiceImpl).RequestTimeout = 100 * time.Millisecond
-		defer func() {
-			th.App.HTTPService().(*httpservice.HTTPServiceImpl).RequestTimeout = httpservice.RequestTimeout
-		}()
+		th.App.UpdateConfig(func(cfg *model.Config) {
+			cfg.ServiceSettings.OutgoingIntegrationRequestsTimeout = model.NewInt64(1)
```
Comment (pvev) [inline server/channels/app/slashcommands/command_test.go:435]: q: is there a reason why we don't have  `model.NewInt8` for these small numbers?

File: server/channels/app/integration_action.go [hunk 6]
```diff
@@ -40,12 +41,7 @@ import (
 	"github.com/mattermost/mattermost/server/v8/channels/utils"
 )
 
-func (a *App) DoPostAction(c *request.Context, postID, actionId, userID, selectedOption string) (string, *model.AppError) {
```
Comment (pvev) [inline server/channels/app/integration_action.go:43]: q: is it intentional the removal of this function? if so, maybe the documentation in the top of this file should be updated then

File: webapp/channels/src/i18n/en.json [hunk 7]
```diff
@@ -2179,6 +2179,8 @@
   "admin.service.iconTitle": "Enable integrations to override profile picture icons:",
   "admin.service.insecureTlsDesc": "When true, any outgoing HTTPS requests will accept unverified, self-signed certificates. For example, outgoing webhooks to a server with a self-signed TLS certificate, using any domain, will be allowed. Note that this makes these connections susceptible to man-in-the-middle attacks.",
   "admin.service.insecureTlsTitle": "Enable Insecure Outgoing Connections: ",
+  "admin.service.integrationRequestDesc": "The number of seconds to wait for Integration requests. That includes <slashCommands>Slash Commands</slashCommands>, <outgoingWebhooks>Outgoing Webhooks</outgoingWebhooks>, <interactiveMessages>Interative Messages</interactiveMessages> and <interactiveDialogs>Interative Dialogs</interactiveDialogs>.",
```
Comment (pvev) [inline webapp/channels/src/i18n/en.json:2182]: there is a small typo in this line: `Interative` instead of `Interactive`

**Review by pvev — APPROVED** [review]
LGTM! thanks @hanzei! :shipit:

File: e2e-tests/playwright/support/server/default_config.ts (hunk 2 above)
Comment (hanzei) [reply e2e-tests/playwright/support/server/default_config.ts:94]: 3/5 that the godoc comment already make it clear that it's in seconds. I'm happy to also add it to the webapp files.

File: server/i18n/en.json (hunk 4 above)
Comment (hanzei) [reply server/i18n/en.json:8012]: I don't quite get what you are referring to. `Duration` does not need to get translated. The translator only needs to put the variable in the right place.

File: server/channels/app/webhook_test.go (hunk 3 above)
Comment (hanzei) [reply server/channels/app/webhook_test.go]: edde00a8dd

File: server/channels/app/integration_action.go (hunk 6 above)
Comment (hanzei) [reply server/channels/app/integration_action.go:43]: e43bca73a5

Comment (hanzei) [reply server/channels/app/integration_action.go:43]: Yes, the removal is intentional. `DoPostAction` is not longer called in the codebase.

File: webapp/channels/src/i18n/en.json (hunk 7 above)
Comment (hanzei) [reply webapp/channels/src/i18n/en.json]: 2ae6ea7e77

Comment (hanzei) [reply webapp/channels/src/i18n/en.json]: Good catch :+1:

File: server/channels/app/slashcommands/command_test.go (hunk 5 above)
Comment (hanzei) [reply server/channels/app/slashcommands/command_test.go:435]: The reason an `int64` is used is that `time.Duration` is an `int64`, and hence it's easier to multiply.

**Comment by hanzei** [comment]
@mickmister I would appreciate a high-level review from you once you are back.

File: server/i18n/en.json (hunk 4 above)
Comment (isacikgoz) [reply server/i18n/en.json:8012]: Thanks, that's what I was asking.

**Review by isacikgoz — APPROVED** [review]

File: server/channels/api4/integration_action_test.go [hunk 8]
```diff
@@ -175,41 +176,62 @@ func TestOpenDialog(t *testing.T) {
...
-	request.Dialog.Elements = []model.DialogElement{}
-	_, err = client.OpenInteractiveDialog(context.Background(), request)
-	require.NoError(t, err)
+	t.Run("Should pass with empty elements slice", func(t *testing.T) {
+		request.Dialog.Elements = []model.DialogElement{}
+		_, err := client.OpenInteractiveDialog(context.Background(), request)
+		require.NoError(t, err)
+	})
+
+	t.Run("Should fail if trigger timeout is extended", func(t *testing.T) {
```
Comment (mickmister) [inline server/channels/api4/integration_action_test.go:224]:
```suggestion
	t.Run("Should fail if trigger timeout is exceeded", func(t *testing.T) {
```

**Review by mickmister — APPROVED** [review]
LGTM

**Comment by guneshsji** [comment]
@hanzei 
I believe remaining reviewer is AFK, 
Is it possible for you to push the changes without them?

**Comment by hanzei** [comment]
@DHaussermann I would like to get these changes into 8.1. Would you please give the PR a review or ask another QA tester for one?

**Comment by DHaussermann** [comment]
@hanzei This is likely something I'd need t take on. I'm unclear on the scope. Can I please get a bit of guidance on this if possible?

- Please confirm this will not impact exiting plugins that have timeouts set a the code level? (possibly a silly question) 
- Is there a relatively quick way to setup testable version of Custom slash commands and Post actions? 
- Can I use exiting plugins like Demo or GitHub to test post action?
- How can I intentionally throttle the response times to hit the timeout? This doesn't feel like something I can do by slowing down client side bandwidth.

**Comment by hanzei** [comment]
1. The stricter timeout always wins. If a plugin sets a shorter timeout for its own request handling, that one will cancel the request first.
2. Good question. I don't think there is a way to test this change without code changes. In my testing, I modified the demo plugin to take longer to respond.
3. Yes, Interactive dialog requests can also be tested this way.
4. As pointed out in 2, this can be done via code changes. Using infrastructure to test that seems more complicated to me.

**Comment by DHaussermann** [comment]
@hanzei I will make time for this PR this week as I don't think there would be a more suitable QA resource short of finding a dev to peer test it. 

On that note can you share how I can modify the demo plugin locally so I can use it to validate this change?

**Comment by hanzei** [comment]
@DHaussermann Thanks, Dylan!

If you insert a `time.Sleep(5*time.Second) between https://github.com/mattermost/mattermost-plugin-demo/blob/a2691134448d402e147cc95ebf7e6e7c6925d70a/server/command_hooks.go#L198-L199, the slash command execution will be delayed by 5 seconds allowing you to test the timeout.

**Comment by hanzei** [comment]
@DHaussermann What are the next steps to getting this merged?

**Comment by hanzei** [comment]
@DHaussermann Gentle ping

**Comment by DHaussermann** [comment]
Hi @hanzei, sorry for the delays on this. I have tried a couple times to get this looked at but it's time consuming for me as there is no obvious way to test this. 

For each of the affected areas I need a delay that I know will reliably exceed the delay set in the config.

So far as discussed above, I have modified the DemoPlugin code to insert a delay which could conceivably cover the changes to slash commands more broadly. But I don't quite see this working as expected. This make me wondering if the test approach is valid and if trying to insert similar delays into other plugins for interactive messages and interactive dialogues is plausible. 

I also still need a way to test outgoing webhooks.

Is it possibel we need some type of simple app to test this?

I short, I'm realizing I need some test steps in order to get this done in a reasonable amount of effort.

Can you please DM me to discuss the details of this?

**Comment by hanzei** [comment]
@DHaussermann I've submitted a PR to the demo plugin that allows you to configure a delay: https://github.com/mattermost/mattermost-plugin-demo/pull/167. I hope this helps. If not, please let me know.

**Comment by DHaussermann** [comment]
Using the updated version of the Demo-Plugin I have now tested Interactive dialogues and slash commands and I see them working as expected. The request can be made to exceed the `Integration request timeout` in system console and fail expectedly. I can also increase the time out to see them succeed with the desired delay 👍 

@hanzei I have 1 question here. Do we set the out-of-the-box value here to 30 seconds? Is that really what we want? When I play with a current server version, it seems like this timeout happens after about 3 seconds. This seems like a large jump once clients upgrade their server version.  

Also, Do you have any thoughts on outgoing webhook or Post Actions? 
For post actions - is it possible to modify a plugin like MS calendar locally to a dd a delay as it use this for the "settings panel"

**Comment by hanzei** [comment]
> @hanzei I have 1 question here. Do we set the out-of-the-box value here to 30 seconds? Is that really what we want? When I play with a current server version, it seems like this timeout happens after about 3 seconds. This seems like a large jump once clients upgrade their server version.

I think the default value makes sense. The current timeout of 3 seconds for opening interactive dialogues is quite low. That has caused various issues in the past. By using a longer timeout, we provide an out-of-the-box experience that is fitting for most use cases.

> Also, Do you have any thoughts on outgoing webhook or Post Actions? For post actions - is it possible to modify a plugin like MS calendar locally to a dd a delay as it use this for the "settings panel"

Sure, we can do that. But that brings us back to the original issue that you could not test the PR without custom builds. Are you fine with me providing such a custom build?

**Comment by hanzei** [comment]
@larkox The test you added in https://github.com/mattermost/mattermost/issues/24602 seems to be flaky. It did fail on this PR.

**Comment by larkox** [comment]
@hanzei Thanks for the heads up. I think I know where the issue might be.

It seems that the order in which the plugin hooks are called is not deterministic in the test. So I have to make the test more deterministic. Will create a PR.

**Comment by DHaussermann** [comment]
@hanzei it would be ideal if the demo plugin could provide post actions for updating token values or some such thing. Is this something that can easily be implemented?

Also in the interest of getting this PR merged - Do you have any thoughts on how to test Outgoing webhooks?

**Comment by hanzei** [comment]
@DHaussermann You can actually test post actions via the `/ephemeral` command. That one creates a new ephemeral post that uses post actions.

**Comment by hanzei** [comment]
@DHaussermann For outgoing webhooks, I'm not quite sure. I _could_ extend the Demo plugin to accept an outgoing webhook, but outgoing webhooks are not really a plugin think. It would make the demo plugin more of a test plugin. @mickmister What do you think?

**Comment by mickmister** [comment]
> For outgoing webhooks, I'm not quite sure. I _could_ extend the Demo plugin to accept an outgoing webhook, but outgoing webhooks are not really a plugin think. It would make the demo plugin more of a test plugin. @mickmister What do you think?

It is a bit odd, but I think it's probably the best option. Otherwise we would instead create a separate HTTP server for the purpose of testing this. 1/5 Might as well have all of the functionality this implements to be tested in one spot. We can put a comment in the handler saying that this is just for testing, as are the other changes for testing this PR

**Comment by hanzei** [comment]
@DHaussermann I would like to get the PR into v9.3 if possible. Would you mind taking a look at this PR and demo plugin onehttps://github.com/mattermost/mattermost-plugin-demo/pull/171 ?

**Comment by amyblais** [comment]
We're at feature complete and no longer taking features to v9.3

**Comment by amyblais** [comment]
@DHaussermann @hanzei A kind note that v9.4 feature complete is on Dec 8th if we want this included in v9.4.

**Comment by DHaussermann** [comment]
@hanzei once this is available on Docker I'll test it again. I know have a working outgoing webhook setup so I can test this regardless on the demo plugin PR. But, I'll look over the logging you added. This may help when I'm trying to make the one contained within the demo plugin work.  👍

**Review by DHaussermann — APPROVED** [review]
Tested and passed

Using https://github.com/mattermost/mattermost-plugin-demo/pull/171 I can confirm that the Integration Management timeout value is also functional for outgoing webhooks. 

This last test concludes the affected areas for this new config. 

LGTM!

Huge thanks for all your help (and patience) on this @hanzei 🎉  much appreciated!

**Pull request merged**